src/couch_replicator.erl

```diff
@@ -137,49 +148,96 @@ cancel_replication(RepId, #user_ctx{name = Name, roles = Roles}) ->
+stream_terminal_docs_info(Db, Cb, UserAcc, States) ->
     DDoc = <<"_replicator">>,
     View = <<"terminal_states">>,
     QueryCb = fun handle_replicator_doc_query/2,
     Args = #mrargs{view_type = map, reduce = false},
-    Acc = {Db, Cb, UserAcc},
+    Acc = {Db, Cb, UserAcc, States},
     try fabric:query_view(Db, DDoc, View, QueryCb, Acc, Args) of
-    {ok, {Db, Cb, UserAcc1}} ->
+    {ok, {Db, Cb, UserAcc1, States}} ->
         UserAcc1
     catch error:database_does_not_exist ->
         UserAcc
     end.
 
 
-% Streaming doesn't save memory much here but it is used to provided a similar
-% API as the terminal docs, which is streaming data from the view
--spec stream_active_docs_info(user_doc_cb(), any()) -> any().
-stream_active_docs_info(Cb, UserAcc) ->
-    {Replies, _BadNodes} = rpc:multicall(couch_replicator_doc_processor, docs, []),
-    lists:foldl(Cb, UserAcc, lists:append(Replies)).
+-spec stream_active_docs_info(user_doc_cb(), any(), [atom()]) -> any().
+stream_active_docs_info(Cb, UserAcc, States) ->
+    Nodes = lists:usort([node() | nodes()]),
+    stream_active_docs_info(Nodes, Cb, UserAcc, States).
+
+
+-spec stream_active_docs_info([node()], user_doc_cb(), any(), [atom()]) -> any().
+stream_active_docs_info([], _Cb, UserAcc, _States) ->
+    UserAcc;
+
+stream_active_docs_info([Node | Nodes], Cb, UserAcc, States) ->
+    case rpc:call(Node, couch_replicator_doc_processor, docs, [States]) of
+        {badrpc, Reason} ->
+            ErrMsg = "Could not get replicator docs from ~p. Error: ~p",
+            couch_log:error(ErrMsg, [Node, Reason]),
+            stream_active_docs_info(Nodes, Cb, UserAcc, States);
+        Results ->
+            UserAcc1 = lists:foldl(Cb, UserAcc, Results),
+            stream_active_docs_info(Nodes, Cb, UserAcc1, States)
+    end.
 
 
 -spec handle_replicator_doc_query
     ({row, [_]} , query_acc()) -> {ok, query_acc()};
     ({error, any()}, query_acc()) -> {error, any()};
     ({meta, any()}, query_acc()) -> {ok,  query_acc()};
     (complete, query_acc()) -> {ok, query_acc()}.
-handle_replicator_doc_query({row, Props}, {Db, Cb, UserAcc}) ->
+handle_replicator_doc_query({row, Props}, {Db, Cb, UserAcc, States}) ->
     DocId = couch_util:get_value(id, Props),
-    State = couch_util:get_value(key, Props),
+    DocStateBin = couch_util:get_value(key, Props),
+    DocState = try
+        erlang:binary_to_existing_atom(DocStateBin, utf8)
+    catch
+        error:badarg ->
+            unknown
+    end,
     StateInfo = couch_util:get_value(value, Props),
-    EjsonInfo = {[
-        {doc_id, DocId},
-        {database, Db},
-        {id, null},
-        {state, State},
-        {info, StateInfo}
-    ]},
-    {ok, {Db, Cb, Cb(EjsonInfo, UserAcc)}};
+    % Set the error_count to 1 if failed. This is mainly for consistency as
+    % jobs from doc_processor and scheduler will have that value set
+    ErrorCount = case DocState =:= failed of
+        true ->
```

Comment:
It's my stylistic preference that for small/simple case statements, you squish them down to four lines. E.g.

```
case Blah of
    foo -> 4;
    bar -> 99
end
```